sbi/src/api/tsm.rs

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::TeeFunction::*; // TODO
+use crate::{ecall_send, Error, Result, SbiMessage};
+use crate::{TsmInfo, TsmPageType, TvmCpuRegister, TvmCreateParams}; //TODO
+
+/// Initiates a TSM fence on this CPU.
+pub fn initiate_fence() -> Result<()> {
+    let msg = SbiMessage::Tee(TsmInitiateFence);
+    // Safety: TsmInitiateFence doesn't read or write any memory we have access to.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Returns information about the TSM that booted this host context.
+pub fn get_info() -> Result<TsmInfo> {
+    let mut tsm_info = TsmInfo::default();
+    let tsm_info_size = core::mem::size_of::<TsmInfo>() as u64;
+    let msg = SbiMessage::Tee(TsmGetInfo {
+        dest_addr: &mut tsm_info as *mut _ as u64,
+        len: tsm_info_size,
+    });
+    // Safety: The passed info pointer is uniquely owned so it's safe to modify in SBI.
+    let tsm_info_len = unsafe { ecall_send(&msg).expect("TsmGetInfo failed") };
+
+    if tsm_info_len != tsm_info_size {
+        return Err(Error::Failed);
+    }
+
+    Ok(tsm_info)
+}
+
+/// Converts the given page range to confidential memory for use in creating or filling pages of a
+/// TVM.
+///
+/// # Safety
+///
+/// The address provided must point to memory that won't be accessed again by the calling program
+/// until it is reclaimed from confidential memory.
+pub unsafe fn convert_pages(addr: u64, num_pages: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TsmConvertPages {
+        page_addr: addr,
+        page_type: TsmPageType::Page4k,
+        num_pages,
+    });
+    // Safety: The passed-in pages are unmapped and we do not access them again until they're
+    // reclaimed.
+    ecall_send(&msg)?;
+
+    // Fence the pages we just converted.
+    initiate_fence()
```

Comment:
Not sure we want to imply `initate_fence()` in convert pages, e.g. if we want to batch convert a bunch of pages first before fencing. Also we probably should have a wrapper for `local_fence()`.

Reply:
for sure, I meant to remove that, done.